Create Forgot Password Endpoint (#45)

Buh-La-Kay opened the pull request:
### Description
Implemented forgot_password and reset_password endpoints as well as a EmailSender class.

### Solution
forgot_password endpoint takes a JSON package containing an email as a PUT operation, responds with an OK (200) status, and throws a UserNotFoundException if there is no user with the provided email. It then creates a random temporary 32 character token which is assigned to the user in the database and appended to a password reset link. Upon following the link the user is brought to the password_reset endpoint. The password_reset endpoint takes the token appended to the url as well as a JSON package containing the new password as a PUT operation, responds with an OK (200) status, and throws a UserNotFoundException if the token is not found. If a user with the token is found, that users password is updated and the token is deleted.

Also implemented an EmailSender class. There is a JavaMailSender bean created in authorization/src/main/java/com/terabite/authorization/service/EmailConfiguration.java that is configured to send from a gmail account that contains two environment variables for login and password. These environment variables will need to be created by individual team members on their local machine if they want to send emails. As of right now we do not have a client email to use so I have been using my personal gmail account to test. Other services that want to use the EmailSender class should just need to make a sendXYZEmail method that sets the recipient address, email header, and email body and then call sendEmail(email, subject, body) with their sendXYZEmail method.

### Testing
Does the project build?
Yes

What else did you do to test this?
Tested the forgot_password endpoint with emails not in the database and it threw the expected exception. Tested the reset_password endpoint with bad tokens and it threw the expected exception.

ZhouDanny reviewed (commented):
For the compose.yaml is there a reason why you changed the ports of 3306:3306 to '3306'? As for the MYSQL_DATABASE 'db' whichever one works is fine, I used db mainly because I rather type 2 letters as opposed to however long 'mydatabase' is.

ZhouDanny reviewed (approved):
As for the email sending stuff I saw you show it to me earlier so no concerns. And I looked over the implementation of the reset_password endpoint. My only concern is how we are using loginRepository, I'm not knowledgeable about what that class does if it stores all the database login as Java objects. But can we instead switch to a static approach if that is not the case?

Buh-La-Kay commented:
> As for the email sending stuff I saw you show it to me earlier so no concerns. And I looked over the implementation of the reset_password endpoint. My only concern is how we are using loginRepository, I'm not knowledgeable about what that class does if it stores all the database login as Java objects. But can we instead switch to a static approach if that is not the case?

I don't think it necessarily stores the login database as Java objects, but it is our bridge to interact with the database through Spring. Since LoginRepository extends JpaRepository spring automatically creates a bean for it on program start. That means that if we want to interact with the "Login" database we have to do it through the singleton "LoginRepository" object that spring automatically creates for us.

Sanguine-Choleric reviewed (changes requested):
Mostly fine, but still a couple of changes/fixes I'd like. Most things in this review are just sanity checks.

* `User.class` and `UserInformation.class` look like duplicates
* Make sure all your errors are handled. I don't know best practice here, but maybe unit test them? I don't know which ones are breaking/non-breaking, so it's your call here.
* Make sure you're not committing dotfiles/IDE config files

Buh-La-Kay commented:
>     * `User.class` and `UserInformation.class` look like duplicates
> 
>     * Make sure you're not committing dotfiles/IDE config files

I removed User class since it was doing nothing, thanks for bringing attention to it. Also added .vscode to my gitignore, thanks for pointing that out as well.

Buh-La-Kay commented on authorization/compose.yaml:
Changed ports back to how it was. As far as the database name I was under the impression we were changing that as a team due to discord discussions, I will change it back if that is not the case.

Sanguine-Choleric replied:
Yea, db name doesn't really matter and can be whatever.

Sanguine-Choleric replied:
We need this file btw, we can't create docker containers w/out it.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/model/User.java:
```diff
@@ -0,0 +1,97 @@
+    private String sex;
+
+
+    //private String shirt_size;
```
Class removed

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/LoginService.java:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    public void updatePassword(Login login, String newPassword){
+
+        //we should have password encoding here
+        login.setPassword(newPassword);
+
+        login.setResetPasswordToken(null);
+        loginRepository.save(login);
+    }
```
Was not being used, removed.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/UserService.java:
No it is not, removed.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/LoginService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void updatePasswordResetToken(String token, String email) throws UserNotFoundException {
+        
+        Login login=loginRepository.findByEmail(email);
```
If I understand how this works correctly (I'd say I'm roughly 75% sure), I believe login will just get assigned to null if there is not a matching email in the login database so there shouldn't be any way of it crashing here. I tried refactoring this to be in a try catch block and got an "unreachable catch block" error.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/controller/AuthorizationController.java:
```diff
@@ -34,4 +56,34 @@ public Payload userLoginPost() {
         return new Payload("Reached logout POST");
     }
+    
+    @PutMapping("/forgot_password")
+    @ResponseStatus(HttpStatus.OK)
+    public void processForgotPassword(@RequestBody String jsonEmail) throws UnsupportedEncodingException, MessagingException, JsonProcessingException, UserNotFoundException{
+        ObjectMapper objectMapper = new ObjectMapper();
+        Email email=objectMapper.readValue(jsonEmail, Email.class);
+        Login login=loginRepository.findByEmail(email.getEmail());
+        if(login!= null){
+            emailSender.sendForgotPasswordEmail(email.getEmail());
+        }
+        else{
+            throw new UserNotFoundException("No user with email "+email.getEmail());
+        } 
+    }
+
+    @PutMapping("/reset_password")
+    @ResponseStatus(HttpStatus.OK)
+    public void resetPassword(@RequestParam String token, @RequestBody String jsonPassword) throws JsonProcessingException, UserNotFoundException{
+        ObjectMapper objectMapper = new ObjectMapper();
+        Password password=objectMapper.readValue(jsonPassword, Password.class);
+        Login login=loginRepository.findByPasswordResetToken(token);
+        if(login==null){
+            throw new UserNotFoundException("Incorrect Password Reset Token");
+        }
+        else{
+            login.setPassword(password.getPassword());
+            login.setResetPasswordToken(null);
+            loginRepository.save(login);
+        }
+    }
```
Created a ForgotPasswordHelper class to process forgot_password and reset_password endpoints. Business logic no longer in controller.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/EmailSender.java:
```diff
@@ -0,0 +1,67 @@
+        //This will change with jwt tokens
+        //random 32 character string
+        String token = UUID.randomUUID().toString();
+        loginRepository.findByEmail(email);
+        loginService.updatePasswordResetToken(token, email);
```
Added try/catch to lines 51/52 as well as try/catch above in sendEmail

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/controller/AuthorizationController.java:
```diff
@@ -15,13 +30,20 @@
 @RequestMapping("/v1/user")
 public class AuthorizationController {
     @Autowired
-    private UserRepository memberRepository;
+    private UserRepository userRepository;
+
+    @Autowired
+    private LoginRepository loginRepository;
+    
+    @Autowired
+    private EmailSender emailSender;
```
Changed all instances of field injection to dependency injection. Much safer in terms of causing errors and should help us down the line with testing.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/EmailSender.java:
```diff
@@ -0,0 +1,67 @@
+@Service
+public class EmailSender {
+
+    @Autowired
+    private JavaMailSender mailSender;
+
+    @Autowired
+    private LoginService loginService;
+
+    @Autowired
+    private LoginRepository loginRepository;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private UserRepository uRepository;
```
Changed to dependency injection

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/LoginService.java:
```diff
@@ -0,0 +1,46 @@
+@Transactional
+public class LoginService {
+    
+    @Autowired
+    private LoginRepository loginRepository;
```
Changed to dependency injection

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/repository/LoginRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+
+public interface LoginRepository extends JpaRepository<Login, Long> {
+    Login findByEmail(String email);
+
+    Login findByPasswordResetToken(String passwordResetToken);
```
In my testing it seems like instead of crashing the program these simply return void if the target is not found. That being said, I still have error catching everywhere these are used just in case.

Sanguine-Choleric reviewed (commented):
Changes are good. I'll test a little more, then approve

Sanguine-Choleric commented on authorization/src/main/java/com/terabite/authorization/controller/AuthorizationController.java:
```diff
@@ -34,4 +41,16 @@ public Payload userLoginPost() {
         return new Payload("Reached logout POST");
     }
+    
+    @PutMapping("/forgot_password")
+    @ResponseStatus(HttpStatus.OK)
+    public void forgotPassword(@RequestBody String jsonEmail) {
+        forgotPasswordHelper.processForgotPassword(jsonEmail);
+    }
```
This endpoint returns code 200 OK even if email to check is invalid. Errors should be transparent so the frontend guys know something went wrong. You should probably just send 400 Bad Request or other relevant response code + a json body with something like "status: success/failed".

Buh-La-Kay replied:
Fixed to return 400 with a bad request and 200 with a good one as well as a message. ("User found/ not found")

Sanguine-Choleric commented on authorization/src/main/java/com/terabite/authorization/controller/AuthorizationController.java:
```diff
@@ -34,4 +41,16 @@ public Payload userLoginPost() {
+        forgotPasswordHelper.processForgotPassword(jsonEmail);
+    }
+
+    @PutMapping("/reset_password")
+    @ResponseStatus(HttpStatus.OK)
+    public void resetPassword(@RequestParam String token, @RequestBody String jsonPassword) throws JsonProcessingException, LoginNotFoundException{
+        forgotPasswordHelper.processResetPassword(token, jsonPassword);
+    }
```
Same thing as above. Make it clear when something goes wrong so it's easier for frontend to work with this.

Buh-La-Kay replied:
Done, 400 for bad request, 200 for good with a message for both.

Sanguine-Choleric reviewed (changes requested):
Just some clarifying changes to make stuff easier for frontend. I'm assuming that the email-sending part works fine because I can't test that.

Buh-La-Kay commented on authorization/src/main/java/com/terabite/authorization/service/Password.java:
No it is not. Misunderstood ObjectMapper requirements

Sanguine-Choleric commented on containers/mysql/Dockerfile:
Docker stuff in `containers/mysql` isn't being used, but I'm just making a note here for future us that we intentionally deleted it

Sanguine-Choleric commented on containers/mysql/docker-compose.yml:
Docker stuff in `containers/mysql` isn't being used, but I'm just making a note here for future us that we intentionally deleted it

Sanguine-Choleric commented on containers/mysql/schema.sql:
Docker stuff in `containers/mysql` isn't being used, but I'm just making a note here for future us that we intentionally deleted it

Sanguine-Choleric commented on authorization/src/main/java/com/terabite/authorization/service/ForgotPasswordHelper.java:
Looks good. I'd personally have the return body be json for consistency, but current return is good enough. Up to you to keep or change to json.

Sanguine-Choleric commented on .gitignore:
We may need to keep this. Doesn't affect backend, dunno about frontend

Sanguine-Choleric reviewed (changes requested):
Final review I think. 
* Restore some key files that you removed - `authorization/compose.yaml` and `authorization/.gitignore`
* Optionally have your `ForgotPasswordHelper.java` functions return json in their body

Sanguine-Choleric reviewed (approved):
LGTM. Just keep in mind how you're gonna test email stuff when we have to start unit testing
